src/kakarot/state.cairo

```diff
@@ -0,0 +1,506 @@
+// SPDX-License-Identifier: MIT
+
+%lang starknet
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import HashBuiltin, BitwiseBuiltin
+from starkware.cairo.common.dict import dict_read, dict_write
+from starkware.cairo.common.default_dict import default_dict_new, default_dict_finalize
+from starkware.cairo.common.dict_access import DictAccess
+from starkware.cairo.common.memcpy import memcpy
+from starkware.cairo.common.registers import get_fp_and_pc
+from starkware.cairo.common.uint256 import Uint256
+from starkware.starknet.common.syscalls import emit_event
+from starkware.cairo.common.uint256 import uint256_add, uint256_sub
+from starkware.starknet.common.storage import normalize_address
+from starkware.cairo.common.hash_state import hash_finalize, hash_init, hash_update, hash_felts
+
+from kakarot.accounts.library import Accounts
+from kakarot.interfaces.interfaces import IAccount, IContractAccount, IERC20
+from kakarot.model import model
+from kakarot.constants import native_token_address, contract_account_class_hash
+from starkware.starknet.common.syscalls import call_contract
+
+namespace State {
+    // @dev Create a new empty State
+    func init() -> model.State* {
+        let (accounts_start) = default_dict_new(0);
+        let (balances_start) = default_dict_new(0);
+        let (events: model.Event*) = alloc();
+        let (transfers: model.Transfer*) = alloc();
+        return new model.State(
+            accounts_start=accounts_start,
+            accounts=accounts_start,
+            events_len=0,
+            events=events,
+            balances_start=balances_start,
+            balances=balances_start,
+            transfers_len=0,
+            transfers=transfers,
+        );
+    }
+
+    // @dev Deep copy of the state, creating new memory segments
+    // @param self The pointer to the State
+    func copy{range_check_ptr}(self: model.State*) -> model.State* {
+        alloc_locals;
+        let self = finalize(self);
```

Comment:
it also copies the events and transfer lists, otherwise if during an eventually reverted tx events/transfers are appended, they would be appened to the parent state

Reply:
true but do we need to squash at every sub context finalisation?
i'm not against but i'm not sure i understand why we do it